hugo/content/docs/recipes/scoping/file-based.md

```diff
@@ -0,0 +1,200 @@
+        ;
+    }
+}
+```
+
+After that, you need to register the `HelloWorldScopeComputation` in the `HelloWorldModule`:
+
+```typescript
+export const HelloWorldModule: Module<HelloWorldServices, PartialLangiumServices & HelloWorldAddedServices> = {
+    //...
+    references: {
+        ScopeComputation: (services) => new HelloWorldScopeComputation(services)
+    }
+};
+```
+
+Having done this, will make all persons that are marked with the `export` keyword available to the other files through the index manager.
+
+## Step 3: Bending the cross-reference resolution
+
+The final step is to adjust the cross-reference resolution through overriding the `DefaultScopeProvider.getScope(…)` function. Here is the implementation:
+
+```typescript
+export class HelloWorldScopeProvider extends DefaultScopeProvider {
+    override getScope(context: ReferenceInfo): Scope {
+        switch(context.container.$type as keyof HelloWorldAstType) {
+            case 'PersonImport':
+                if(context.property === 'person') {
+                    return this.getExportedPersonsFromGlobalScope(context);
+                }
+                break;
+            case 'Greeting':
+                if(context.property === 'person') {
+                    return this.getImportedPersonsFromCurrentFile(context);
+                }
+                break;
+        }
+        return EMPTY_SCOPE;
+    }
+    //...
+}
+```
+
+Do not forget to add the new service to the `HelloWorldModule`:
+
+```typescript
+export const HelloWorldModule: Module<HelloWorldServices, PartialLangiumServices & HelloWorldAddedServices> = {
+    //...
+    references: {
+        ScopeComputation: (services) => new HelloWorldScopeComputation(services),
+        ScopeProvider: (services) => new HelloWorldScopeProvider(services) //NEW!
+    }
+};
+```
+
+You noticed the two missing functions? Here is what they have to do.
+
+The first function (`getExportedPersonsFromGlobalScope(context)`) will take a look at the global scope and return all exported persons respecting the files that were touched by the file imports. Note that we are outputting all persons that are marked with the `export` keyword. The actual name resolution is done internally later by the linker.
+
+```typescript
+protected getExportedPersonsFromGlobalScope(context: ReferenceInfo): Scope {
+    //get document for current reference
+    const document = AstUtils.getDocument(context.container);
+    //get model of document
+    const model = document.parseResult.value as Model;
+    //determine current directory
+    const currentDir = dirname(document.uri.fsPath);
+    //look at all imports of the current document
+    const astNodeDescriptions = model.fileImports.flatMap((fileImport) => {
+        const fileName = resolve(currentDir, fileImport.file);
+        const uri = URI.file(fileName);
+        return this.indexManager.allElements(Person, new Set<string>([uri.toString()])).toArray();
+    });
```

Comment:
Suggestion: We can do this using a single `indexManager#allElements` call. Also using `file` here makes this unavailable for non-file scheme URIs

```suggestion
    const currentUri = document.uri;
    const currentDir = dirname(currentUri.path);
    const uris = new Set<string>();
    for (const fileImport of model.fileImports) {
      const filePath = join(currentDir, fileImport.file);
      const uri = currentUri.with({ path: filePath });
      uris.add(uri.toString());
    }
    const astNodeDescriptions = this.indexManager.allElements(Person, uris).toArray();
```

Reply:
Added also comments